compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1565,6 +1565,9 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                             && let hir::ExprKind::Loop(loop_blk, ..) = expression.kind {
                               intravisit::walk_block(& mut visitor, loop_blk);
                         }
+                        if let Some(expr) = expression {
+                            self.note_result_coercion(fcx, &mut err, expr, expected, found);
```

Comment:
Ideally we'd put this on `FnCtxt` and add it to `emit_type_mismatch_suggestions` too, for non-`CoerceMany` type mismatches.

Reply:
Done. I do have some concerns about this triggering when it shouldn't, like in `let _: Result<(), A> = Err(B);` telling you to write `Ok(Err(B)?)`, which wouldn't work. (It doesn't, but I don't know if it is only because of a lower priority.)